sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala

```diff
@@ -272,6 +275,83 @@ class ParquetFileFormat
     true
   }
 
+  override def getSplits(fileIndex: FileIndex,
+                         fileStatus: FileStatus,
+                         filters: Seq[Filter],
+                         schema: StructType,
+                         hadoopConf: Configuration): Seq[FileSplit] = {
+    if (filters.isEmpty) {
+      // Return immediately to save FileSystem overhead
+      super.getSplits(fileIndex, fileStatus, filters, schema, hadoopConf)
+    } else {
+      val filePath = fileStatus.getPath
+      val rootOption: Option[Path] = fileIndex.rootPaths
+        .find(root => filePath.toString.startsWith(root.toString))
+      val metadataOption = rootOption.flatMap(getMetadataForPath(filePath, _, hadoopConf))
+      // If the metadata exists, filter the splits.
+      // Otherwise, fall back to the default implementation.
+      metadataOption
+        .map(filterToSplits(fileStatus, _, rootOption.get, filters, schema, hadoopConf))
+        .getOrElse(super.getSplits(fileIndex, fileStatus, filters, schema, hadoopConf))
+    }
+  }
+
+  private def filterToSplits(fileStatus: FileStatus,
+                             metadata: ParquetMetadata,
+                             metadataRoot: Path,
+                             filters: Seq[Filter],
+                             schema: StructType,
+                             hadoopConf: Configuration): Seq[FileSplit] = {
+    val metadataBlocks = metadata.getBlocks
+    val parquetSchema = metadata.getFileMetaData.getSchema
+    val filter = FilterCompat.get(filters
+          .flatMap(ParquetFilters.createFilter(schema, _))
+          .reduce(FilterApi.and))
+    val filteredMetadata =
+      RowGroupFilter.filterRowGroups(filter, metadataBlocks, parquetSchema).asScala
+    filteredMetadata.flatMap(bmd => {
+      val bmdPath = new Path(metadataRoot, bmd.getPath)
+      val fsPath = fileStatus.getPath
+      if (bmdPath == fsPath) {
+        Some(new FileSplit(bmdPath, bmd.getStartingPos, bmd.getTotalByteSize, Array.empty))
+      } else {
+        None
+      }
+    })
+  }
+
+  private def getMetadataForPath(filePath: Path,
```

Comment:
I think this should be cacheable, as the lifetime of a `FileFormat` is tied to a particular `HadoopFsRelation`, which again is tied to a `Dataset`. Something may have changed about the API but from a quick look through the relevant classes I don't see anything to refute this. I don't see any written guarantees in the FileFormat API that concrete impls need to be stateless, but neither did I find any reference to caching/state in other implementations. Still I think it should be safe to cache on the ParquetFileFormat for this use case.